src/main/java/com/fasterxml/jackson/databind/introspect/AnnotatedCreatorCollector.java

```diff
@@ -241,14 +248,129 @@ Creators collect(Class<?> primaryMixIn)
                 result.set(i,
-                        constructFactoryCreator(candidates.get(i),
+                        constructFactoryCreator(candidate,
                                 typeResCtxt, null));
             }
         }
         return result;
     }
 
+    /*
+     * Attempt to narrow types on a generic factory method based on the expected result (_primaryType).
+     * If narrowing was possible, a new TypeResolutionContext is returned with the discovered TypeBindings,
+     * otherwise the emptyTypeResCtxt argument is returned.
+     *
+     * For example:
+     * Given type Wrapper<T> with
+     * @JsonCreator static <T> Wrapper<T> fromJson(T value)
+     * When a Wrapper<Duck> is requested the factory must return a Wrapper<Duck> and we can bind T to Duck.
+     */
+    private TypeResolutionContext narrowMethodGenericTypeArguments(
+            TypeResolutionContext emptyTypeResCtxt, Method candidate) {
+        TypeVariable<Method>[] methodTypeParameters = candidate.getTypeParameters();
+        if (methodTypeParameters.length == 0
+                // If the primary type has no type parameters, there's nothing to do
+                || _primaryType.getBindings().isEmpty()) {
+            // Method has no type parameters: no need to modify the resolution context.
+            return emptyTypeResCtxt;
+        }
+        Type genericReturnType = candidate.getGenericReturnType();
+        if (!(genericReturnType instanceof ParameterizedType)) {
+            // Return value is not parameterized, it cannot be used to associate the _primaryType expectations
+            // onto parameters.
+            return emptyTypeResCtxt;
+        }
+
+        ParameterizedType parameterizedGenericReturnType = (ParameterizedType) genericReturnType;
+        // Primary type and result type must be the same class, otherwise we would need to
+        // trace generic parameters to a common superclass or interface.
+        if (!Objects.equals(_primaryType.getRawClass(), parameterizedGenericReturnType.getRawType())) {
+            return emptyTypeResCtxt;
+        }
+
+        // Construct a mapping from the method type variable names to the types requested by _primaryType
+        // For example given _primaryType: Foo<String, Int>
+        // and method static <T, U> Foo<T, U> func(Object in)
+        // This produces {T=String, U=Int}.
+        Type[] methodReturnTypeArguments = parameterizedGenericReturnType.getActualTypeArguments();
+        Map<String, JavaType> bindTargetsByName = new HashMap<>(methodReturnTypeArguments.length);
+        for (int j = 0; j < methodReturnTypeArguments.length; j++) {
+            Type methodReturnTypeArgument = methodReturnTypeArguments[j];
+            // Note: This strictly supports only TypeVariables, not wildcards with nested type variables
+            if (methodReturnTypeArgument instanceof TypeVariable) {
```

Comment:
I'll just note that this class ideally should not do anything with JDK's non-Class type constructs -- this is what `TypeFactory` is intended to encapsulate. I realize that occasionally workarounds may be needed but this is something I feel strongly about and is based on work writing `java-classmate` (improved version of generic type resolution based on Jackson's internals).

Reply:
I imagine this can be ported to JavaType+TypeFactory easily enough, we may lose some ability to detect edge cases, but that's probably a reasonable trade-off. We can try it out and see how it impacts tests.

Would you like me to make this change?